scene/2d/tile_map_layer.cpp

```diff
@@ -2565,6 +2657,105 @@ void TileMapLayer::set_enabled(bool p_enabled) {
 bool TileMapLayer::is_enabled() const {
 	return enabled;
 }
+void TileMapLayer::set_tileset(const Ref<TileSet> &p_tileset) {
+	if (p_tileset == tile_set) {
+		return;
+	}
+
+	dirty.flags[DIRTY_FLAGS_TILE_SET] = true;
+	_queue_internal_update();
+	emit_signal(CoreStringNames::get_singleton()->changed);
+
+	// Set the tileset, registering to its changes.
+	if (tile_set.is_valid()) {
+		tile_set->disconnect_changed(callable_mp(this, &TileMapLayer::_tile_set_changed));
+	}
+
+	tile_set = p_tileset;
+
+	if (tile_set.is_valid()) {
+		tile_set->connect_changed(callable_mp(this, &TileMapLayer::_tile_set_changed));
+	}
+
+	update_configuration_warnings();
+}
+
+Ref<TileSet> TileMapLayer::get_tileset() const {
+	return tile_set;
+}
+
+void TileMapLayer::set_tile_map_data_from_array(const Vector<uint8_t> &p_data) {
+	const int cell_data_struct_size = 12;
+
+	int size = p_data.size();
+	const uint8_t *ptr = p_data.ptr();
+
+	// Index in the array.
+	int index = 0;
+
+	// First extract the data version.
+	int16_t format = decode_uint16(&ptr[index]);
+	index += 2;
+	ERR_FAIL_COND_MSG(format != TileMapLayerDataFormat::TILE_MAP_LAYER_DATA_FORMAT_MAX - 1, vformat("Unsupported tile map data format: %s. Expected: %s", format, TileMapLayerDataFormat::TILE_MAP_LAYER_DATA_FORMAT_MAX - 1));
```

Comment:
Shouldn't this check for `format > FORMAT_MAX`? If it's lower, it can be handled with compatibility code.

Reply:
No. I took the opportunity to reset the "format" to 0 while creating the new TileMapLayer node. So there's no backward compatibility to be taken care of here (it's done in the TileMap node instead).

There's now a `TileMapLayerDataFormat` and a `TileMapDataFormat` now :)

Reply:
I mean, if the format changes in the future, the condition will need to be changed anyway. Unless you plan to never allow loading older formats.